src/single.rs

```diff
@@ -335,126 +310,113 @@ fn serde_error_from_group_decoding_error<ERR: ::serde::de::Error>(err: GroupDeco
-    /// or `[u8; 96]` which satisfy `pairing::EncodedPoint` and permit
-    /// creation and write access with `pairing::EncodedPoint::empty()`
-    /// and `AsMef<[u8]>`, respectively.
-    pub fn decompress(compressed: <<<E as EngineBLS>::$group as CurveProjective>::Affine as CurveAffine>::Compressed) -> Result<Self,GroupDecodingError> {
-        Ok($wrapper(compressed.into_affine()?.into_projective()))
-    }
-
-    pub fn decompress_from_slice(slice: &[u8]) -> Result<Self,GroupDecodingError> {
-        let mut compressed = <<<E as EngineBLS>::$group as CurveProjective>::Affine as CurveAffine>::Compressed::empty();
-        if slice.len() != compressed.as_mut().len() {
-            // We should ideally return our own error here, but this seems acceptable for now.
-            return Err(GroupDecodingError::UnexpectedInformation);
-        } // <<<E as EngineBLS>::$group as CurveProjective>::Affine as CurveAffine>::Compressed::size() 
-        compressed.as_mut().copy_from_slice(slice);
-        $wrapper::<E>::decompress(compressed)
-    }
-}
-
-#[cfg(feature = "serde")]
-impl<E> ::serde::Serialize for $wrapper<E> where E: $se {
-    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: ::serde::Serializer {
-        serializer.serialize_bytes(self.compress().as_ref())
-    }
-}
-
-#[cfg(feature = "serde")]
-impl<'d,E> ::serde::Deserialize<'d> for $wrapper<E> where E: $de {
-    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: ::serde::Deserializer<'d> {
-        use std::fmt;
-        use std::marker::PhantomData;
-
-        struct MyVisitor<EE: $de>(PhantomData<EE>);
-
-        impl<'d,EE: $de> ::serde::de::Visitor<'d> for MyVisitor<EE> {
-            type Value = $wrapper<EE>;
-
-            fn expecting(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
-                formatter.write_str(Self::Value::DESCRIPTION)
+        impl<E> CanonicalSerialize for $wrapper<E> where E: $se {
+            #[inline]
+            fn serialize<W: Write>(&self, mut writer: W) -> Result<(), SerializationError> {
+                self.0.into_affine().serialize(&mut writer)?;
+                Ok(())
             }
-
-            fn visit_bytes<ERR>(self, bytes: &[u8]) -> Result<$wrapper<EE>, ERR> where ERR: ::serde::de::Error {
-                $wrapper::<EE>::decompress_from_slice(bytes)
-                .map_err(serde_error_from_group_decoding_error)
+            #[inline]
+            fn serialized_size(&self) -> usize {
+                self.0.into_affine().serialized_size()
+            }    
+            #[inline]
+            fn serialize_uncompressed<W: Write>(&self, mut writer: W) -> Result<(), SerializationError> {
+                self.0.into_affine().serialize_uncompressed(&mut writer)?;
+                Ok(())
+            }
+            #[inline]
+            fn uncompressed_size(&self) -> usize {
+                self.0.into_affine().uncompressed_size()
+            }
```

Comment:
Can we just pass this through to the `CanonicalSerialize` on the underlying type?

I'm not sure what I think of this `CanonicalSerialize` trait overall yet, but I guess it's what arkworks does.  Is it what celo does?

Reply:
hmm, we can't. because projective points can't be canonically serialized (and hence `ProjectiveCurve` trait does not implement `CanonicalSerialize`. I think turn them into affine and call the serialization functions is the closest thing we do. I can look into celo's approach though.

Reply:
Are we serializing projective without going through affine here?  Why?  Did i do that?

Reply:
No, we go through affine as in `self.0.into_affine()` . What I meant was that we just can't simply derive `CanonicalSerialize` for the Public/Private key because we need to perform this `into_affine()` step manually. 

You basically reimplemented serialization from scratch, there was no official serialization in the original pairing library.

Reply:
We could pass this through to the underlying affine curve then maybe?

Reply:
I'll just merge this and then we can make another issue for passing these through.

Reply:
> We could pass this through to the underlying affine curve then maybe?

I thought I was doing that. Could you show me what is the proper way of doing it?